rust/relay/ebpf-turn-router/src/move_payload.rs

```diff
@@ -0,0 +1,306 @@
+use aya_ebpf::{helpers::bpf_xdp_adjust_tail, programs::XdpContext};
+use network_types::{
+    eth::EthHdr,
+    ip::{Ipv4Hdr, Ipv6Hdr},
+    udp::UdpHdr,
+};
+
+use crate::{channel_data::CdHdr, error::Error};
+
+// Set an upper limit for bounds checks
+const MAX_MTU: usize = 1500;
+const MAX_PAYLOAD: usize = MAX_MTU - EthHdr::LEN - Ipv4Hdr::LEN - UdpHdr::LEN - CdHdr::LEN;
+
+#[inline(always)]
+pub fn add_channel_data_header_ipv4(ctx: &XdpContext, header: CdHdr) -> Result<(), Error> {
+    extend_and_shift_backward::<{ Ipv4Hdr::LEN }>(ctx, &header)
+}
+
+#[inline(always)]
+pub fn add_channel_data_header_ipv6(ctx: &XdpContext, header: CdHdr) -> Result<(), Error> {
+    extend_and_shift_backward::<{ Ipv6Hdr::LEN }>(ctx, &header)
+}
+
+#[inline(always)]
+pub fn remove_channel_data_header_ipv4(ctx: &XdpContext) -> Result<(), Error> {
+    shift_forward_and_shrink::<{ Ipv4Hdr::LEN }>(ctx)
+}
+
+#[inline(always)]
+pub fn remove_channel_data_header_ipv6(ctx: &XdpContext) -> Result<(), Error> {
+    shift_forward_and_shrink::<{ Ipv6Hdr::LEN }>(ctx)
+}
+
+/// Extend tail by channel data header length and write channel data header at the front of the
+/// payload.
+#[inline(never)]
+fn extend_and_shift_backward<const IP_HEADER_LEN: usize>(
+    ctx: &XdpContext,
+    header: &CdHdr,
+) -> Result<(), Error> {
+    let off = payload_offset::<IP_HEADER_LEN>();
+
+    // Bounds check for verifier
+    if ctx.data() + off > ctx.data_end() {
+        return Err(Error::PacketTooShort);
+    }
+
+    // Extend tail
+    let ret = unsafe { bpf_xdp_adjust_tail(ctx.ctx, CdHdr::LEN as i32) };
+    if ret < 0 {
+        return Err(Error::XdpAdjustTailFailed(ret));
+    }
+
+    let data_size = ctx.data_end() - ctx.data();
+
+    // Copy payload
+    let original_data_size = data_size - CdHdr::LEN;
+    let payload_size = original_data_size - off;
+    let src_off = original_data_size;
```

Comment:
If you inline `original_data_size` here, you can see the symmetry with the other function better.

Reply:
Yeah I thought there was probably an optimization to make here. Let me give it a shot.